kernel/popcorn/syscall_server.c

```diff
@@ -0,0 +1,229 @@
+DEFINE_SYSCALL_REDIRECT(epoll_pwait, PCN_SYSCALL_EPOLL_PWAIT,int, epfd,
+			struct epoll_event __user *, events, int, maxevents,
+			int, timeout, const sigset_t __user *, sigmask,
+			size_t, sigsetsize);
+DEFINE_SYSCALL_REDIRECT(epoll_ctl, PCN_SYSCALL_EPOLL_CTL, int, epfd,
+			int, op, int, fd, struct epoll_event __user *,
+			event);
+
+
+/* General fs/driver read/write/open/close calls */
+DEFINE_SYSCALL_REDIRECT(read, PCN_SYSCALL_READ, unsigned int, fd, char __user*,
+			buf, size_t, count);
+DEFINE_SYSCALL_REDIRECT(write, PCN_SYSCALL_WRITE, unsigned int, fd, const char
+			__user *, buf, size_t, count);
+DEFINE_SYSCALL_REDIRECT(open, PCN_SYSCALL_OPEN, const char __user *, filename,
+			int, flags, umode_t, mode);
+DEFINE_SYSCALL_REDIRECT(close, PCN_SYSCALL_CLOSE, unsigned int, fd);
+DEFINE_SYSCALL_REDIRECT(ioctl, PCN_SYSCALL_IOCTL, unsigned int, fd,
+			unsigned int, cmd, unsigned long, arg);
+DEFINE_SYSCALL_REDIRECT(writev, PCN_SYSCALL_WRITEV, unsigned long,
+			fd, const struct iovec __user *, vec,
+			unsigned long, vlen);
+DEFINE_SYSCALL_REDIRECT(fstat, PCN_SYSCALL_FSTAT, unsigned int, fd,
+			struct stat __user *, statbuf);
+DEFINE_SYSCALL_REDIRECT(sendfile64, PCN_SYSCALL_SENDFILE64,int, out_fd, int,
+			in_fd, loff_t __user *, offset, size_t, count);
+/**
+ * Syscalls needed in the kernel
+ * */
+extern int sys_socket(int family, int type, int protocol);
+extern int sys_bind(int fd, struct sockaddr __user *umyaddr, int addrlen);
+extern int sys_listen(int fd, int backlog);
+extern int sys_accept4(int fd, struct sockaddr __user *upeer_sockaddr,
+		     int __user *upeer_addrlen, int flag);
+extern int sys_setsockopt(int fd, int level, int optname, char __user *optval,
+			  int optlen);
+extern long sys_recvfrom(int, void __user *, size_t, unsigned,
+				struct sockaddr __user *, int __user *);
+extern long sys_shutdown(int, int);
+extern long sys_epoll_create1(int flags);
+extern long sys_epoll_ctl(int epfd, int op, int fd,
+				struct epoll_event __user *event);
+extern long sys_epoll_wait(int epfd, struct epoll_event __user *events,
+				int maxevents, int timeout);
+extern long sys_epoll_pwait(int epfd, struct epoll_event __user *events,
+				int maxevents, int timeout,
+				const sigset_t __user *sigmask,
+				size_t sigsetsize);
+extern long sys_read(unsigned int fd, char __user *buf, size_t count);
+extern long sys_write(unsigned int fd, const char __user *buf, size_t count);
+extern long sys_open(const char __user *filename, int flags, umode_t mode);
+extern long sys_close(unsigned int fd);
+extern long sys_ioctl(unsigned int fd, unsigned int cmd, unsigned long arg);
+extern long sys_writev(unsigned long fd,
+			   const struct iovec __user *vec,
+			   unsigned long vlen);
+extern long sys_newfstat(unsigned int fd, struct stat __user *statbuf);
+extern long sys_sendfile64(int out_fd, int in_fd,
+			       loff_t __user *offset, size_t count);
+
+int process_remote_syscall(struct pcn_kmsg_message *msg)
```

Comment:
Please add the comment for this function. Also please follow the kernel comment style.
https://www.kernel.org/doc/html/v4.10/process/coding-style.html#commenting